### PR TITLE
[WIP] Add API for comment reactions

### LINKED ISSUE #8313: API: handle reactions to comments
```
GET     /repos/{owner}/{repo}/issues/comments/{id}/reactions
PUT     /repos/{owner}/{repo}/issues/comments/{id}/reactions
DELETE  /repos/{owner}/{repo}/issues/comments/{id}/reactions
```

### PR DESCRIPTION
**Pull request by 6543**
close #8313

### CODE REVIEW

**Comment by 6543** [comment]
only one problem is bloking the WIP state:

if you do `http://localhost:3000/api/v1/repos/user1/repo1/issues/comments/4/reactions`:

```
reflect: NumField of non-struct type
/usr/lib/go/src/reflect/type.go:981 (0x48a13d)
	(*rtype).NumField: panic("reflect: NumField of non-struct type")
/home/6543/git/own/gitea/vendor/gitea.com/macaron/binding/binding.go:597 (0x106dbdb)
	mapForm: for i := 0; i < typ.NumField(); i++ {
/home/6543/git/own/gitea/vendor/gitea.com/macaron/binding/binding.go:149 (0x107074d)
	Form.func1: errors = mapForm(formStruct, ctx.Req.Form, nil, errors)
/usr/lib/go/src/reflect/value.go:460 (0x491cc5)
	Value.call: call(frametype, fn, args, uint32(frametype.size), uint32(retOffset))
/usr/lib/go/src/reflect/value.go:321 (0x491483)
	Value.Call: return v.call("Call", in)
/home/6543/git/own/gitea/vendor/gitea.com/macaron/inject/inject.go:177 (0x9a7249)
	(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil
/home/6543/git/own/gitea/vendor/gitea.com/macaron/inject/inject.go:137 (0x9a6bf9)
	(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())
/home/6543/git/own/gitea/vendor/gitea.com/macaron/binding/binding.go:63 (0x106898e)
	bind: ctx.Invoke(Form(obj, ifacePtr...))
/home/6543/git/own/gitea/vendor/gitea.com/macaron/binding/binding.go:107 (0x1070475)
	Bind.func1: bind(ctx, obj, ifacePtr...)
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/context.go:79 (0x9d6c40)
	ContextInvoker.Invoke: invoke(params[0].(*Context))
/home/6543/git/own/gitea/vendor/gitea.com/macaron/inject/inject.go:157 (0x9a6f59)
	(*injector).fastInvoke: return f.Invoke(in)
/home/6543/git/own/gitea/vendor/gitea.com/macaron/inject/inject.go:135 (0x9a6ce8)
	(*injector).Invoke: return inj.fastInvoke(v, t, t.NumIn())
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/context.go:121 (0x9d6d98)
	(*Context).run: vals, err := c.Invoke(c.handler())
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/context.go:112 (0x1095801)
	(*Context).Next: c.run()
/home/6543/git/own/gitea/modules/context/panic.go:39 (0x10957ed)
	Recovery.func1: ctx.Next()
/usr/lib/go/src/reflect/value.go:460 (0x491cc5)
	Value.call: call(frametype, fn, args, uint32(frametype.size), uint32(retOffset))
/usr/lib/go/src/reflect/value.go:321 (0x491483)
	Value.Call: return v.call("Call", in)
/home/6543/git/own/gitea/vendor/gitea.com/macaron/inject/inject.go:177 (0x9a7249)
	(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil
/home/6543/git/own/gitea/vendor/gitea.com/macaron/inject/inject.go:137 (0x9a6bf9)
	(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/context.go:121 (0x9d6d98)
	(*Context).run: vals, err := c.Invoke(c.handler())
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/context.go:112 (0xa6e9c5)
	(*Context).Next: c.run()
/home/6543/git/own/gitea/vendor/gitea.com/macaron/session/session.go:192 (0xa6e9b0)
	Sessioner.func1: ctx.Next()
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/context.go:79 (0x9d6c40)
	ContextInvoker.Invoke: invoke(params[0].(*Context))
/home/6543/git/own/gitea/vendor/gitea.com/macaron/inject/inject.go:157 (0x9a6f59)
	(*injector).fastInvoke: return f.Invoke(in)
/home/6543/git/own/gitea/vendor/gitea.com/macaron/inject/inject.go:135 (0x9a6ce8)
	(*injector).Invoke: return inj.fastInvoke(v, t, t.NumIn())
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/context.go:121 (0x9d6d98)
	(*Context).run: vals, err := c.Invoke(c.handler())
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/context.go:112 (0x9e8439)
	(*Context).Next: c.run()
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/recovery.go:161 (0x9e8427)
	Recovery.func1: c.Next()
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/logger.go:40 (0x9da913)
	LoggerInvoker.Invoke: invoke(params[0].(*Context), params[1].(*log.Logger))
/home/6543/git/own/gitea/vendor/gitea.com/macaron/inject/inject.go:157 (0x9a6f59)
	(*injector).fastInvoke: return f.Invoke(in)
/home/6543/git/own/gitea/vendor/gitea.com/macaron/inject/inject.go:135 (0x9a6ce8)
	(*injector).Invoke: return inj.fastInvoke(v, t, t.NumIn())
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/context.go:121 (0x9d6d98)
	(*Context).run: vals, err := c.Invoke(c.handler())
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/context.go:112 (0x9e7770)
	(*Context).Next: c.run()
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/logger.go:52 (0x9e775b)
	Logger.func1: ctx.Next()
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/logger.go:40 (0x9da913)
	LoggerInvoker.Invoke: invoke(params[0].(*Context), params[1].(*log.Logger))
/home/6543/git/own/gitea/vendor/gitea.com/macaron/inject/inject.go:157 (0x9a6f59)
	(*injector).fastInvoke: return f.Invoke(in)
/home/6543/git/own/gitea/vendor/gitea.com/macaron/inject/inject.go:135 (0x9a6ce8)
	(*injector).Invoke: return inj.fastInvoke(v, t, t.NumIn())
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/context.go:121 (0x9d6d98)
	(*Context).run: vals, err := c.Invoke(c.handler())
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/router.go:187 (0x9e9656)
	(*Router).Handle.func1: c.run()
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/router.go:303 (0x9e3035)
	(*Router).ServeHTTP: h(rw, req, p)
/home/6543/git/own/gitea/vendor/gitea.com/macaron/macaron/macaron.go:220 (0x9dbcaa)
	(*Macaron).ServeHTTP: m.Router.ServeHTTP(rw, req)
/home/6543/git/own/gitea/vendor/github.com/gorilla/context/context.go:141 (0xc2840d)
	ClearHandler.func1: h.ServeHTTP(w, r)
/usr/lib/go/src/net/http/server.go:2007 (0x752a03)
	HandlerFunc.ServeHTTP: f(w, r)
/usr/lib/go/src/net/http/server.go:2802 (0x755e53)
	serverHandler.ServeHTTP: handler.ServeHTTP(rw, req)
/usr/lib/go/src/net/http/server.go:1890 (0x7517f4)
	(*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)
/usr/lib/go/src/runtime/asm_amd64.s:1357 (0x461090)
	goexit: BYTE	$0x90	// NOP
```

**Comment by jolheiser** [comment]
@6543 
I believe it's because you made a type definition, not a struct. 

```go
type CommentReactionList []*CommentReaction
```

I _think_ what you want is 
```go
type CommentReacionList struct {
    CommentReactions []*CommentReaction `json:"comment_reactions"`
}
```
or something similar

Reflect is mad because a type definition reflection doesn't have a `NumField`

**Comment by 6543** [comment]
@jolheiser thanks this fixes the reflection ...

now i only get
`422  Undocumented | Error: Unprocessable Entity`

Response body:
```
[
  {
    "fieldNames": [
      "CommentReactions"
    ],
    "classification": "RequiredError",
    "message": "Required"
  }
]
```

File: routers/api/v1/repo/issue_comment.go [hunk 1]
```diff
@@ -393,3 +393,183 @@ func deleteIssueComment(ctx *context.APIContext) {
...
+	comment, err := models.GetCommentByID(ctx.ParamsInt64(":id"))
+	if err != nil {
+		if models.IsErrCommentNotExist(err) {
+			ctx.NotFound(err)
+		} else {
+			ctx.Error(500, "GetCommentByID", err)
+		}
+		return
+	}
+	issue, err := models.GetIssueByIndex(ctx.Repo.Repository.ID, comment.IssueID)
```
Comment (lunny) [inline routers/api/v1/repo/issue_comment.go]: `IssueID` Or `IssueIndex`?

File: models/issue_reaction.go [hunk 2]
```diff
@@ -247,3 +255,24 @@ func (list ReactionList) GetMoreUserCount() int {
...
 }
+
+// APIFormat returns Raction in api Format
+func (list ReactionList) APIFormat() api.CommentReactionList {
+	result := []*api.CommentReaction{}
+	users := make(map[string][]string)
+	counts := make(map[string]int64)
+
+	for _, r := range list {
+		users[r.Type] = append(users[r.Type], r.User.LoginName)
```
Comment (lunny) [inline models/issue_reaction.go]: r.User should be loaded before this invoked.

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/issue_reaction.go (hunk 2 above)
Comment (6543) [reply models/issue_reaction.go]: I'll do `u := r.User` now
is this what you ment?

**Comment by 6543** [comment]
because there were no "reactions" anymore ... I'll implement the API like described in https://github.com/go-gitea/gitea/issues/8313#issuecomment-550568899 ... hopefully this is good enouth

**Comment by 6543** [comment]
#9220 replace this PR

**Pull request closed without merging**